lib/spack/spack/cmd/location.py

```diff
@@ -69,7 +68,7 @@ def location(parser, args):
     elif args.env:
         path = spack.environment.root(args.env)
         if not os.path.isdir(path):
-            tty.die("no such environment: '%s'" % args.env)
+            raise BadEnvironmentError("No such environment: '%s'" % args.env)
```

Comment:
I know I made a comment [here](https://github.com/spack/spack/pull/11663#discussion_r291787811) suggesting that we should replace `tty.die` invocations with `SpackError` subclasses, but I meant for *core code*.  This code is in a command -- where it's fine (and probably clearer) to use `tty.die()` instead of introducing many different exception types.  Commands are really what `tty.die()` is for.  You can think of them as just an extended `main()` method, where calling `exit()` is fine.  They're not library code - they're supposed to be simple, and if things get complex in the commands, we try to push the complexity into the core so that their logic stays simple.

The only thing that invokes a command besides the framework in `spack.main` is tests, and for *command* tests, it's sufficient to just do `with pytest.raises(SystemExit):`.  The one I complained about [here](https://github.com/spack/spack/pull/11663#discussion_r291787811) was a method of `DIYStage` (a core class) calling `tty.die()`.  Sorry for confusing things.

Basically, I don't think we need to go overboard on exception types here.  My general rule of thumb is that we shouldn't define new exception types in the commands, because it clutters up the command logic for not a lot of gain.  I'd just revert this change and the others in `cmd/location.py`.

Reply:
I suspected as such wrt `tty.die` versus exceptions but it's good to get confirmation.

Reply:
Thankfully I found the `fail_on_error` option so I could still check the unique output message instead of `SpackCommandError`s standard `Command exited with code 1: location(<args>)`.